### PR TITLE
Cbo 1 anti join heuristics

### PR DESCRIPTION
**Pull request by sopel39**
This causes outer join stats to be slightly overestimated.

I have checked that q78 works as fast in cbo_automatic as ecj_repartitioned (around 24 secs).
ecj_automatic works faster (<20secs).

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/cost/JoinStatsRule.java [hunk 1]
```diff
@@ -49,12 +49,21 @@
         implements ComposableStatsCalculator.Rule
 {
     private static final Pattern PATTERN = Pattern.typeOf(JoinNode.class);
+    private static final double DEFAULT_UNMATCHED_ANTI_JOIN_NDVS_COEFFICIENT = 0.5;
```
Comment (cawallin) [inline presto-main/src/main/java/com/facebook/presto/cost/JoinStatsRule.java]: Why .5? I'd imagine it should be similar to the UNKNOWN_FILTER_COEFFICIENT, which is .1.

**Comment by losipiuk** [comment]
Travis failures.

File: presto-main/src/main/java/com/facebook/presto/cost/JoinStatsRule.java [hunk 2]
```diff
@@ -331,11 +340,13 @@ private PlanNodeStatsEstimate crossJoinStats(JoinNode node, PlanNodeStatsEstimat
     {
         private final PlanNodeStatsEstimate stats;
         private final Optional<Symbol> selectedClauseSymbol;
+        private final double selectedClauseSymbolNdv;
 
-        public AntiJoinStats(PlanNodeStatsEstimate stats, Optional<Symbol> selectedClauseSymbol)
+        public AntiJoinStats(PlanNodeStatsEstimate stats, Optional<Symbol> selectedClauseSymbol, double selectedClauseSymbolNdv)
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/cost/JoinStatsRule.java]: It is hard to follow. 
It looks like the commit mixes two things. 
1. Adding `DEFAULT_UNMATCHED_ANTI_JOIN_NDVS_COEFFICIENT`
2. refactor to add `selectdClauseSymbolNdv` to `AntiJoinStats`

I do not understand why we want `2`. It makes `AntiJoinStats` a junk bag. It does not seem to fit in here.

File: presto-main/src/main/java/com/facebook/presto/cost/JoinStatsRule.java (hunk 1 above)
Comment (sopel39) [reply presto-main/src/main/java/com/facebook/presto/cost/JoinStatsRule.java]: This is not for unknown values. This is for know columns, but we have to assume that not every NDV (from join side with less NDVs) matches "the other side". I tested and 0.5 causes slight overestimation (especially in q78) in `left joins`, which is probably good. I am not sure thought that the value is the best possible. If someone figures that some other value is better that is fine for me.
Lower value would increase anti join result set.
Greater values would decrease anti join result set.

File: presto-main/src/main/java/com/facebook/presto/cost/JoinStatsRule.java (hunk 2 above)
Comment (sopel39) [reply presto-main/src/main/java/com/facebook/presto/cost/JoinStatsRule.java]: `selectdClauseSymbolNdv` is required in order to cap NDVs when adding inner join to anti join stats. It is now possible that inner NDVs + anti join NDVs would be larger than original column NDVs.
I will probably just pass max NDV explicitly

File: presto-main/src/main/java/com/facebook/presto/cost/JoinStatsRule.java (hunk 1 above)
Comment (cawallin) [reply presto-main/src/main/java/com/facebook/presto/cost/JoinStatsRule.java]: What value would make most sense with our otherwise assumption of uniform distribution?

q05, q40, q49, q72, q75, q77, q78, q80, and q93 are the queries in TPC-DS that have LEFT JOINs, we should benchmark those in particular with and without your change

Comment (sopel39) [reply presto-main/src/main/java/com/facebook/presto/cost/JoinStatsRule.java]: yeah, we need to ensure there is no regression (or not too large)

**Comment by sopel39** [comment]
applied comemnts

File: presto-main/src/test/java/com/facebook/presto/cost/StatsCalculatorAssertion.java [hunk 3]
```diff
@@ -96,6 +97,14 @@ public StatsCalculatorAssertion check(Consumer<PlanNodeStatsAssertion> statistic
         return this;
     }
 
+    public StatsCalculatorAssertion check(ComposableStatsCalculator.Rule rule, Consumer<PlanNodeStatsAssertion> statisticsAssertionConsumer)
```
Comment (losipiuk) [inline presto-main/src/test/java/com/facebook/presto/cost/StatsCalculatorAssertion.java:100]: I suggest that you add `forRule` method which will set `Optional<Rule>` within `StatsCalculatorAssertion`. And then when stats are computed and `rule` is set then it is used for computation instead whole calculator.

File: presto-main/src/main/java/com/facebook/presto/cost/JoinStatsRule.java [hunk 4]
```diff
@@ -257,52 +271,52 @@ else if (leftNDV <= rightNDV) {
...
             result = result.mapOutputRowCount(rowCount -> min(leftStats.getOutputRowCount(), rowCount / UNKNOWN_FILTER_COEFFICIENT));
         }
 
-        return new AntiJoinStats(result, Optional.of(drivingClause.getLeft()));
+        return new JoinComplementStats(result, Optional.of(drivingClause.getLeft()));
     }
 
     @VisibleForTesting
-    PlanNodeStatsEstimate addAntiJoinStats(PlanNodeStatsEstimate joinStats, PlanNodeStatsEstimate antiJoinStats, Optional<Symbol> joinSymbol)
+    PlanNodeStatsEstimate addJoinComplementStats(PlanNodeStatsEstimate innerJoinStats, JoinComplementStats joinComplementStats, double selectedClauseSymbolMaxNdv)
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/cost/JoinStatsRule.java:286]: s/selected/driving

File: presto-main/src/test/java/com/facebook/presto/cost/TestJoinStatsRule.java [hunk 5]
```diff
@@ -258,33 +259,33 @@ public void testStatsForLeftAndRightJoin()
...
         PlanNodeStatsEstimate statsToAdd = planNodeStats(RIGHT_ROWS_COUNT,
                 symbolStatistics(LEFT_JOIN_COLUMN, -5.0, 5.0, 0.2, 5));
 
         PlanNodeStatsEstimate addedStats = planNodeStats(TOTAL_ROWS_COUNT,
-                symbolStatistics(LEFT_JOIN_COLUMN, -5.0, 20.0, (LEFT_ROWS_COUNT * LEFT_JOIN_COLUMN_NULLS + RIGHT_ROWS_COUNT * 0.2) / TOTAL_ROWS_COUNT, 25),
+                symbolStatistics(LEFT_JOIN_COLUMN, -5.0, 20.0, (LEFT_ROWS_COUNT * LEFT_JOIN_COLUMN_NULLS + RIGHT_ROWS_COUNT * 0.2) / TOTAL_ROWS_COUNT, LEFT_JOIN_COLUMN_NDV + 5),
                 symbolStatistics(LEFT_OTHER_COLUMN, 42, 42, (0.42 * LEFT_ROWS_COUNT + RIGHT_ROWS_COUNT) / TOTAL_ROWS_COUNT, 1));
 
-        assertThat(JOIN_STATS_RULE.addAntiJoinStats(LEFT_STATS, statsToAdd, Optional.of((new Symbol(LEFT_JOIN_COLUMN))))).equalTo(addedStats);
+        assertThat(JOIN_STATS_RULE.addJoinComplementStats(LEFT_STATS, new JoinComplementStats(statsToAdd, Optional.of(new Symbol(LEFT_JOIN_COLUMN))), LEFT_JOIN_COLUMN_NDV + 5)).equalTo(addedStats);
```
Comment (losipiuk) [inline presto-main/src/test/java/com/facebook/presto/cost/TestJoinStatsRule.java:288]: I do  not get what `LEFT_JOIN_COLUMN_NDV + 5` mean.

**Review by losipiuk — APPROVED** [review]
Seems ok, but I find the test very unreadable. 
Please give it a thought what to do with it to make it more readable.

File: presto-main/src/main/java/com/facebook/presto/cost/JoinStatsRule.java (hunk 1 above)
Comment (sopel39) [reply presto-main/src/main/java/com/facebook/presto/cost/JoinStatsRule.java]: did some initial benchmarks, there isn't regression (I think), current results q78:
```
ecj repart: 24s
ecj auto: 20s
cbo auto: 28s
```
Is it OK for now?

Comment (cawallin) [reply presto-main/src/main/java/com/facebook/presto/cost/JoinStatsRule.java]: q78 looks fine, and if there isn't a regression for the other queries either then this looks good.
